File: examples/checks/flandor.py
# Independent Python checks for the flandor example.
from .common import run_fragment_checks

CHECKS = [
    ('payload hash matches the source envelope digest', ['Usage is permitted only for purpose "regional_stabilization" and the envelope expires at 2026-04-08T19:00:00+00:00.']),
    ('HMAC value matches the trusted precomputed signature', ['question : Is the Flemish Economic Resilience Board allowed to use a neutral macro-economic insight for regional stabilization, and if so which package should it activate for Fland']),
    ('export weakness, skills strain, and grid stress reach the three-need threshold', ['grid congestion : 19 hours (threshold > 11)']),
    ('the insight has a device scope, event scope, and expiry', ['export weakness : active - at least one cluster has exportOrdersIndex < 90']),
    ('the shared insight omits firm names, payroll rows, and other sensitive terms', ['Payload SHA-256: 718f5b17d07ab6a95503bc04a1000ddb132409f600659c03d21def81914b780b']),
    ('regional stabilization use is authorized before the insight expires', ['Envelope HMAC-SHA-256: 955968ca99a191783bc00cba068128ccb9ff40a5e6114fda13a52c74ee27329e']),
    ('the deletion duty is scheduled before expiry', ['Deletion duty time : 2026-04-08T18:30:00+00:00']),
    ('reuse for firm surveillance is prohibited', ['Surveillance reuse is blocked by a prohibition on odrl:distribute for purpose "firm_surveillance".']),
    ('the recommended package fits inside the €140M budget', ['Budget cap: €140M']),
    ('the recommended package covers all active needs', ['pkg:CORRIDOR_165 : reject, cost=€165M, covers all active needs; over budget by €25M']),
    ('the lowest-cost eligible package is chosen', ['Selected package "Flandor Retooling Pulse" covers export=yes, skills=yes, grid=yes, cost=€120M.']),
    ('the expected six files are recorded as written', ['grid stress : active - grid congestion hours are above 11']),
    ('all source-level policy, scope, signature, and package checks pass', ['pkg:RET_FLEX_120 : selected, cost=€120M, covers all active needs; within budget']),
]

def run(ctx):
    return run_fragment_checks(ctx, CHECKS)

File: examples/checks/flandor.go
package main

import (
	"strings"
)

func checkFlandor(ctx *Context) []Check {
	d := ctx.M()
	needs := map[string]bool{"export": false, "skills": num(asMap(d["Labour"])["TechVacancyRate"]) > 3.9, "grid": num(asMap(d["Grid"])["CongestionHours"]) > 11}
	for _, c := range maps(d["Clusters"]) {
		if num(c["ExportOrdersIndex"]) < 90 {
			needs["export"] = true
		}
	}
	active := 0
	for _, v := range needs {
		if v {
			active++
		}
	}
	covers := func(p anymap) bool {
		return (!needs["export"] || boolean(p["CoversExportWeakness"])) && (!needs["skills"] || boolean(p["CoversSkillsStrain"])) && (!needs["grid"] || boolean(p["CoversGridStress"]))
	}
	var selected anymap
	for _, p := range maps(d["Packages"]) {
		if covers(p) && num(p["CostMEUR"]) <= num(asMap(d["Budget"])["MaxMEUR"]) {
			if selected == nil || num(p["CostMEUR"]) < num(selected["CostMEUR"]) || (num(p["CostMEUR"]) == num(selected["CostMEUR"]) && str(p["PackageID"]) < str(selected["PackageID"])) {
				selected = p
			}
		}
	}
	repActive := grabInt(ctx.Answer, `Active need count: (\d+)/`)
	repPkg := ""
	if m := reFind(ctx.Answer, `Recommended package: ([^\n]+)`); m != nil {
		repPkg = m[0]
	}
	repCost := grabInt(ctx.Answer, `Package cost: €(\d+)M`)
	repHash := ""
	if m := reFind(ctx.Answer, `Payload SHA-256: ([0-9a-f]+)`); m != nil {
		repHash = m[0]
	}
	repHMAC := ""
	if m := reFind(ctx.Answer, `Envelope HMAC-SHA-256: ([0-9a-f]+)`); m != nil {
		repHMAC = m[0]
	}
	insight := asMap(d["Insight"])
	policy := asMap(d["Policy"])
	sig := asMap(d["Signature"])
	allowed := str(d["RequestAction"]) == str(policy["PermissionAction"]) && str(d["RequestPurpose"]) == str(policy["PermissionPurpose"]) && str(policy["PermissionTarget"]) == str(insight["ID"]) && str(d["BoardAuthAt"]) < str(insight["ExpiresAt"]) && active >= integer(insight["ThresholdScore"]) && selected != nil
	cheaperReject := true
	for _, p := range maps(d["Packages"]) {
		if selected != nil && num(p["CostMEUR"]) < num(selected["CostMEUR"]) {
			cheaperReject = cheaperReject && !covers(p)
		}
	}
	full := maps(d["Packages"])[len(maps(d["Packages"]))-1]
	serialized := strings.ToLower(str(insight["SerializedLowercase"]))
	exportWeak := []string{}
	for _, c := range maps(d["Clusters"]) {
		if num(c["ExportOrdersIndex"]) < 90 {
			exportWeak = append(exportWeak, str(c["Name"]))
		}
	}
	return []Check{{"export weakness, skills strain, and grid stress are all active", needs["export"] && needs["skills"] && needs["grid"] && active == 3}, {"active need count meets the insight threshold", repActive == active && active == integer(insight["ThresholdScore"])}, {"the lowest-cost package covering all active needs is selected", str(selected["Name"]) == "Flandor Retooling Pulse" && repPkg == str(selected["Name"])}, {"the selected package fits inside the €140M budget", integer(selected["CostMEUR"]) == repCost && repCost == 120 && num(selected["CostMEUR"]) <= num(asMap(d["Budget"])["MaxMEUR"])}, {"cheaper packages are rejected because each covers only one active need", cheaperReject}, {"the full corridor package covers all needs but is over budget", covers(full) && num(full["CostMEUR"]) > num(asMap(d["Budget"])["MaxMEUR"])}, {"policy permission authorizes regional-stabilization use before expiry", allowed && contains(ctx.Answer, "decision : ALLOWED")}, {"firm-surveillance redistribution is prohibited", str(policy["ProhibitionAction"]) == "odrl:distribute" && str(policy["ProhibitionPurpose"]) == "firm_surveillance"}, {"deletion duty is scheduled before envelope expiry", str(d["BoardDutyAt"]) < str(insight["ExpiresAt"])}, {"shared insight omits firm names and payroll rows", !boolean(asMap(d["Signals"])["ContainsFirmNames"]) && !boolean(asMap(d["Signals"])["ContainsPayrollRows"]) && !contains(serialized, "firm") && !contains(serialized, "payroll")}, {"reported signature metadata matches the trusted precomputed input", repHash == str(sig["DisplayPayloadSHA256"]) && repHMAC == str(sig["SignatureHMAC"]) && str(sig["Algorithm"]) == "HMAC-SHA256"}, {"the expected six files and one audit entry are recorded", integer(d["FilesWritten"]) == integer(d["ExpectedFilesWritten"]) && integer(d["FilesWritten"]) == 6 && integer(d["AuditEntries"]) == 1}, {"export-weak cluster names are independently identified", sliceEq(exportWeak, []string{"Antwerp chemicals", "Ghent manufacturing"})}}
}
